pkg/optimizer/flatzinc_model.go

```diff
@@ -0,0 +1,217 @@
+	goal SolveGoal
+	expr string
+	Annotations
+}
+
+func (slv FlatZincSolveItem) SolveItemStatement() string {
+	return fmt.Sprintf("solve%s %s %s;\n", slv.AnnotationString(), slv.goal, slv.expr)
+}
+
+type FlatZincModel struct {
+	ParamMap    map[string]FlatZincParam
+	VarMap      map[string]FlatZincVariable
+	Constraints []FlatZincConstraint
+	SolveTarget FlatZincSolveItem
+}
+
+func NewFlatZincModel() *FlatZincModel {
+	var fzw FlatZincModel
+	fzw.ParamMap = make(map[string]FlatZincParam)
+	fzw.VarMap = make(map[string]FlatZincVariable)
+	fzw.Constraints = make([]FlatZincConstraint, 0)
+	return &fzw
+}
+
+func (fzw *FlatZincModel) AddParam(numInst uint, name, vartype, assignment string) {
+	fzw.ParamMap[name] = FlatZincParam{numInst, name, vartype, assignment}
+}
+
+func (fzw *FlatZincModel) AddVariable(numInst uint, name, vartype, assignment string, annotations ...string) {
+	fzw.VarMap[name] = FlatZincVariable{numInst, name, vartype, assignment, annotations}
+}
+
+func (fzw *FlatZincModel) AddConstraint(identifier string, exprs []string, annotations ...string) {
+	fzw.Constraints = append(fzw.Constraints, FlatZincConstraint{identifier, exprs, annotations})
+}
+
+func (fzw *FlatZincModel) SetSolveTarget(goal SolveGoal, expr string, annotations ...string) {
+	fzw.SolveTarget = FlatZincSolveItem{goal, expr, annotations}
+}
+
+func (fzw *FlatZincModel) Dump(fileName string) error {
+	file, err := os.Create(path.Clean(fileName))
+	if err != nil {
+		return fmt.Errorf("failed opening file %s for writing", fileName)
+	}
+	defer func() {
+		if err := file.Close(); err != nil {
+			log.Printf("Error closing file: %s\n", err)
+		}
+	}()
+
+	fileContent := ""
+	for _, fzparam := range fzw.ParamMap {
+		fileContent += fzparam.ParamDeclaration()
+	}
+
+	fileContent += "\n"
+	for _, fzvar := range fzw.VarMap {
+		fileContent += fzvar.VarDeclaration()
+	}
+
+	fileContent += "\n"
+	for _, cnstr := range fzw.Constraints {
+		fileContent += cnstr.ConstraintStatement()
+	}
+
+	fileContent += "\n" + fzw.SolveTarget.SolveItemStatement()
+	if _, err := file.WriteString(fileContent); err != nil {
+		return err
+	}
+	return nil
+}
+
+func parseSolutionLine(line string, lineNum uint) (string, []string, error) {
+	err := fmt.Errorf("parse error on line %d: %s", lineNum, line)
+	lineParts := strings.Split(line, "=")
+	if len(lineParts) != 2 {
+		return "", nil, err
+	}
+	varName := lineParts[0]
+	value := strings.TrimSuffix(lineParts[1], ";")
+	values := []string{value}
+	if strings.HasPrefix(value, "array") {
+		leftBracketPos := strings.Index(value, "[")
+		rightBracketPos := strings.Index(value, "]")
+		if leftBracketPos == -1 || rightBracketPos == -1 || leftBracketPos > rightBracketPos {
+			return "", nil, err
+		}
+		values = strings.Split(value[leftBracketPos+1:rightBracketPos], ",")
+	}
+	return varName, values, nil
+}
+
+func (fzw *FlatZincModel) ReadSolutions(fileName string) (map[string][]string, error) {
+	data, err := os.ReadFile(path.Clean(fileName))
```

Comment:
Same question about file based approach

Reply:
Solvers should output their result to `stdout`, so we can read it directly from there without going through files

Reply:
Will this work as expected in case of multiple reconciles simultaneously, i.e. applying the solver in multiple processes with different inputs/outputs?

Reply:
For now, the plan is to block-wait until the CSP solver returns an answer, hoping that it doesn't take too long.
We'll just use `exec.Command("CSPsolver").Output()`. Hence, no issue of mixing answers from different solver runs.